### PR TITLE
ofono QMI: mainline kernel support

### PR DESCRIPTION
**Pull request by scintill**
Tested by @bhush9 on Debian. He sent some SMS!

# TODO before this is completely ready
* [x] write udev rules like described in [rpmsgexport](https://github.com/andersson/rpmsgexport), in a new `modem-qcom-msm-mainline-common` package, which depends on qcom_rmtfs
    * [x] probably need to package rpmsgexport, for above
* [ ] maybe transcribe other channels from [the MSM fork](https://android.googlesource.com/platform/hardware/bsp/kernel/qcom/qcom-msm-v3.10/+/master/drivers/char/msm_smd_pkt.c#846) (see the array above that for /dev names)? Probably not, if we don't have a need for them

Opening this PR might have been premature since I can't test the full polished solution... I didn't want the work to get lost though.

### CODE REVIEW

**Comment by bhush9** [comment]
probably it would be awesome to also include qmimodem call support.. :-)

**Comment by scintill** [comment]
Here's my try at writing udev rules, to automatically create the device node. I'm not very experienced with that, and don't have a mainline device, so it'd be great if someone can test or review.

```
ACTION=="add", SUBSYSTEM=="rpmsg", \
               KERNEL=="rpmsg_ctrl[0-9]*", \
               ATTRS{rpmsg_name}=="modem", \
               RUN+="rpmsgexport /dev/$name DATA5_CNTL"

SUBSYSTEM=="rpmsg", ATTR{rpmsg_name}=="DATA5_CNTL", \
                    NAME="smdcntl0", \
                    ENV{OFONO_DRIVER}="gobi"
```

This is maybe controversial, but I've named the device node smdcntl0, like the MSM fork does. ofono doesn't care about the name, but it might be nice to have consistency for use with path-based tools like qmicli.

**Comment by ollieparanoid** [comment]
Awesome! :tada: Added the `dont_merge_pr` label, as you've listed TODOs that need to be done first.

**Comment by scintill** [comment]
These last changes should finish it. It's rebased onto master. Can @bhush9 or someone with a mainline device try installing `modem-qcom-msm-mainline-common`? (and/or add it as a dependency in your device package.) If it works, after booting you should have a /dev/modem symlink that works with qmicli, and ofono should detect the modem.

File: aports/modem/ofono/APKBUILD [hunk 1]
```diff
@@ -2,7 +2,7 @@
 pkgname=ofono
 _upstreamver=1.21
 pkgver=1.21_p20180307
-pkgrel=0
+pkgrel=2
```
Comment (ollieparanoid) [inline aports/modem/ofono/APKBUILD]: `pkgrel=1`

**Comment by opendata26** [comment]
run command seems wrong gives
[   31.133508] udevd[781]: failed to execute '/lib/udev/rpmsgexport' 'rpmsgexport /dev/rpmsg_ctrl1 DATA5_CNTL': No such file or directory

**Comment by scintill** [comment]
Thank you @opendata26. This should fix that RUN issue. I need to test this merged branch on my device too, but I don't think there should be any issues.

**Comment by scintill** [comment]
It works on my device/kernel.

**Comment by bhush9** [comment]
I think this is good to merge in master, I've changes for hammerhead on top of this PR, @ollieparanoid @scintill Are you fine with merging this PR?

**Comment by ollieparanoid** [comment]
Fine with me as well, what do you think @scintill?

(Maybe @MartijnBraam can merge this one once @scintill confirms?)

**Pull request merged**